chalk-solve/src/coherence.rs

```diff
@@ -72,24 +71,25 @@ impl<I: Interner> SpecializationPriorities<I> {
 #[derive(Copy, Clone, Default, PartialOrd, Ord, PartialEq, Eq, Debug)]
 pub struct SpecializationPriority(usize);
 
-impl<'db, I, S, SC> CoherenceSolver<'db, I, S, SC>
+impl<'db, 'sb, I> CoherenceSolver<'db, 'sb, I>
 where
     I: Interner,
-    S: Solver<I>,
-    SC: Into<S> + Copy,
 {
     /// Constructs a new `CoherenceSolver`.
-    pub fn new(db: &'db dyn RustIrDatabase<I>, solver_choice: SC, trait_id: TraitId<I>) -> Self {
+    pub fn new(
+        db: &'db dyn RustIrDatabase<I>,
+        solver_builder: &'sb dyn Fn() -> Box<dyn Solver<I>>,
+        trait_id: TraitId<I>,
+    ) -> Self {
         Self {
             db,
-            solver_choice,
+            solver_builder,
             trait_id,
-            _solver: std::marker::PhantomData,
         }
     }
 
     pub fn specialization_priorities(
-        &self,
+        &mut self,
```

Comment:
Why do these need `&mut self` now?

Reply:
It's no longer needed now. I have changed them back.